A list of all transactions done by address

Hrushi20 opened the pull request:
* Created a rest api endpoint to fetch all signatures for particular address
* Created Transaction graphql schema to fetch transactions. 


## Todo
* Need to parse transaction data and display on the UI

Hrushi20 committed:
fetch signatures for address, init setup to fetch transactions

Signed-off-by: Hrushi20 <hrushikesh20thegreat@gmail.com>

buffalojoec commented on scripts/generate-transactions.sh:
```diff
@@ -4,4 +4,10 @@ rpc_endpoint=${1:-"http://127.0.0.1:8899"}
\ No newline at end of file
+solana airdrop --url $rpc_endpoint 10
+
+solana transfer CSg2vQGbnwWdSyJpwK4i3qGfB6FebaV3xQTx4U1MbixN 1 --allow-unfunded-recipient --url $rpc_endpoint
```
Solana CLI supports recipients as keypair paths, so what we do here is generate a new mock keypair under `fixtures`, then use that as the recipient.

```
mock_keypair_file="fixtures/mock-keypair.json"

solana-keygen new --no-bip39-passphrase -o $mock_keypair_file
solana transfer --allow-unfunded-recipient --url $rpc_endpoint $mock_keypair_file 1
```

buffalojoec commented on scripts/generate-transactions.sh:
```diff
@@ -4,4 +4,10 @@ rpc_endpoint=${1:-"http://127.0.0.1:8899"}
+
+solana transfer CSg2vQGbnwWdSyJpwK4i3qGfB6FebaV3xQTx4U1MbixN 1 --allow-unfunded-recipient --url $rpc_endpoint
+
+spl_token=$(spl-token create-token --url $rpc_endpoint | awk '{print $3}')
+spl_account_1=$(spl-token create-account $spl_token --url $rpc_endpoint | awk '{print $3}')
+spl-token mint $spl_token 100 --url $rpc_endpoint
```
We can also follow the above approach with the mint and token account creation. We can generate keypairs for stored variables like `mock_mint_keypair_file` and point `spl-token create-token` at those keypair files.

buffalojoec commented on server.ts:
```diff
@@ -33,6 +33,14 @@ app.post('/api/graphql', async (req, res) => {
     return;
 });
 
+app.post('/api/getSignaturesForAddress', async(req, res) => {
+    const { address } = req.body;
+
+    let results = await rpc.getSignaturesForAddress(address).send();
+    const signatures = results.map(result => {return { signature: result.signature }});
+    res.send(JSON.stringify(signatures));
+});
+
```
We should probably use a `try-catch` here to return any errors to the app with a proper server code. I don't use `try-catch` above in the GraphQL response, since GraphQL errors don't throw but are instead returned in the `errors` field.

buffalojoec commented on src/components/Transaction.tsx:
```diff
@@ -15,17 +15,38 @@ const source = /* GraphQL */ `
-            const response = (await gql(source, { signature }));
-            console.log(response)
-            setData(response.account);
+        const fetchSignaturesAndLoadTransactions = async () => {
+            const body = { address: props.address };
+            let signatures = await _fetch('/api/getSignaturesForAddress',body) as Array<Signature>;
+
+            const transactions:Array<Data> = [];
+            signatures.forEach(async (signature) => {
+                const response = (await gql(source, { signature:signature })) as { transaction: Data };
+                console.log(response)
+            })
+
+            setData(transactions);
```
This is in the wrong place. The `Transactions` _page_ should be making this call once, to gather all the signatures for the wallet address, and then using each signature as a property input on a list of `Transaction` components.

buffalojoec commented on src/fetch.ts:
```diff
@@ -20,3 +20,23 @@ export async function gql(...params: Parameters<RpcGraphQL['query']>) {
     });
 }
+
+export async function _fetch(url:string, body: any){
+    const res = await fetch(url, {
+        body: JSON.stringify({
+            ...body
+        }),
+        headers: {
+            'Content-Type': 'application/json',
+        },
+        method: 'POST',
+    });
+
+    const text = await res.text();
+    return JSON.parse(text, (_, value) => {
+        if (value && typeof value === 'object' && '__bigint' in value) {
+            return BigInt(value['__bigint']);
+        }
+        return value;
+    });
+}
```
Since this shares a lot of the same code as the `gql` function above it, maybe we create a function `post` and call into that from `gql`. What do you think?

buffalojoec commented on src/pages/TransactionPage.tsx:
```diff
@@ -1,11 +1,16 @@
         <>
             <h1>This is transaction page</h1>
-            <Transaction/>
+            <Transaction address={LOCAL_WALLET_ADDRESS}/>
```
Reiterating here that we should instead be populating a list of `Transaction` components based on the retrieved signatures for the wallet.

buffalojoec reviewed (commented):
Nice work, left a few comments on possible improvements. Thanks for contributing over here as well!

Hrushi20 committed:
Transaction component, update transaction file, error handling on server

Signed-off-by: Hrushi20 <hrushikesh20thegreat@gmail.com>

Hrushi20 commented:
I updated the code based on comments. I am able to fetch signatures for transactions and updated the script. I also created a Transaction component which is responsible for rendering a `Transaction`. 

The code is not clean. Need to refactor it.

buffalojoec commented on src/pages/TransactionPage.tsx:
```diff
@@ -1,16 +1,37 @@
+    '91RzsA7P8hYDqswUqapHWLSjGiMCFdT3kkoGkcpWZerW' as Address<'91RzsA7P8hYDqswUqapHWLSjGiMCFdT3kkoGkcpWZerW'>;  
 
 export default function TransactionPage() {
+    const [data, setData] = useState<Array<Signature>>(['fdasfsf' as Signature]);
```
This should be initialized empty.

```suggestion
    const [data, setData] = useState<Array<Signature>>();
```

buffalojoec commented on src/pages/TransactionPage.tsx:
```diff
@@ -1,16 +1,37 @@
+                {data.map(signature => <Transaction signature={signature}/>)}
+            </>
+        )
+    }
```
The page itself should be static, and the content within can be dynamic.

```suggestion
  return (
    <>
      <h1> This is transaction page </h1>
      { data
        ? {data.map(signature => <Transaction signature={signature}/>}
        : <h1> Loading transactions... </h1>
      }
    </>
  )
```

buffalojoec commented on src/pages/TransactionPage.tsx:
```diff
@@ -1,16 +1,37 @@
+            // Display on UI that no transactions are found for this address
+            if(signatures){
+                setData(signatures)
+            }
```
If we definitively want a message that says "No transactions found" when the returned list of signatures has length 0, we'll need to handle that case explicitly. You can likely add to my other suggested conditional rendering, something like:

```tsx
  return (
    <>
      <h1> This is transaction page </h1>
      { data
        ? { data.length === 0
              ? <h1> No transactions found. </h1>
              : { data.map(signature => <Transaction signature={signature}/> }
           }
        : <h1> Loading transactions... </h1>
      }
    </>
  )
```

buffalojoec commented on scripts/generate-transactions.sh:
```diff
@@ -6,8 +6,13 @@ echo "Using RPC Endpoint: $rpc_endpoint"
 solana airdrop --url $rpc_endpoint 10
 
-solana transfer CSg2vQGbnwWdSyJpwK4i3qGfB6FebaV3xQTx4U1MbixN 1 --allow-unfunded-recipient --url $rpc_endpoint
+mock_keypair_file="fixtures/mock-keypair.json"
+mock_mint_keypair_file="fixtures/mock-mint-keypair.json"
```
Nice! Sorry, I should have thought of this before, but let's add a command in this script to make a new directory.

```
mkdir mock-keypairs
```

And then let's use that directory, and add it to the `.gitignore` so we don't commit the keypairs.

Also, we'll want to just add the `--force` flag to `solana-keygen new` so we don't have to mess around each time we want to run this.

Hrushi20 replied:
How do I read the address into web app? Do I use the server to use from files for now?

buffalojoec replied:
Let's just copy-paste it for now. We can get to that later.

Hrushi20 replied:
I'm able to fetch transaction data and display it on the UI using graphql.

buffalojoec reviewed (commented):
Shaping up really nicely! Thanks again!

Hrushi20 committed:
create mock-keypair directory, display transaction data

Hrushi20 committed:
replace h3 with p tags, add transcation header

Signed-off-by: Hrushi20 <hrushikesh20thegreat@gmail.com>

buffalojoec commented on server.ts:
```diff
@@ -12,13 +12,13 @@ app.use(bodyParser.json());
     if (response.errors) {
         response.errors.forEach(e => console.error(e));
-        res.status(500).send('Server Error');
+        next({ statusCode:500, message: 'Server Error'});
```
Why change this?

buffalojoec commented on server.ts:
```diff
@@ -33,6 +33,31 @@ app.post('/api/graphql', async (req, res) => {
+}
+
+app.post('/api/getSignaturesForAddress', async(req, res, next) => {
+    try{
```
I think you might need to run the ESLint and Prettier scripts in the `package.json`. There's also a script to auto-format anything that's non-compliant. Mind formatting this PR?

buffalojoec commented on server.ts:
```diff
@@ -33,6 +33,31 @@ app.post('/api/graphql', async (req, res) => {
+        const signatures = results.map(result => result.signature );
+        console.log(signatures)
+        res.status(200).send(JSON.stringify(signatures));
+    }catch(e:any){
+        console.error(e)
+        let error:Error = { statusCode: 500, message: e?.context?.__serverMessage, __code: e?.context?.__code };
+        next(error);
+    }
```
I appreciate the attempt, but since it's only a demo app, you can just log the error and send a 500 like I do with `/graphql`.

```ts
console.error(e);
res.status(500).send('Server Error');
```

buffalojoec commented on server.ts:
```diff
@@ -33,6 +33,31 @@ app.post('/api/graphql', async (req, res) => {
+    }
+});
+
+app.use(function(err:Error, _req:any, res:any, _next:any){
+    res.status(err.statusCode).json({ err });
+});
```
I don't think we need this.

buffalojoec commented on src/components/Transaction.tsx:
```diff
@@ -1,31 +1,112 @@
+
+// Can add more fields based on requirement
 const source = /* GraphQL */ `
+    fragment TransactionMessageAccountKey on TransactionMessageAccountKey {
+        pubkey 
+        signer 
+        source 
+        writable
+    }
```
What's this for?

Hrushi20 replied:
I thought we can show the use of fragments to users. We can remove as it is not used in multiple places.

buffalojoec replied:
Technically there's more to using fragments than just that, but we can save fragments and more GraphQL features for later. Right now let's just lay out the components and pages, and start to tie the app together. Later we can rewire all the queries (using Relay) and optimize this thing.

buffalojoec commented on src/components/Transaction.tsx:
```diff
@@ -1,31 +1,112 @@
+    message: TransactionMessage,
+    // Should have a TransactionMetadata type in sdk
+    meta: {
+        fee: { __bigint: string} | undefined,
```
Fields with `__bigint` should never make it into the app. This is just the mechanism I've employed to send `bigint` values over JSON payloads.

If you look at the server, it will serialize `bigint` values using the `__bigint` key, and then the fetchers in the frontend (`src/fetch.ts`) will deserialize them by converting back to `bigint`.

buffalojoec commented on src/components/Transaction.tsx:
```diff
@@ -1,31 +1,112 @@
+    }
+    signatures: [Signature],
+    slot: Slot,
+    version: String
```
```suggestion
    version: string,
```

buffalojoec commented on src/components/Transaction.tsx:
```diff
@@ -1,31 +1,112 @@
+
+                    <div>
+                        <h3>Transaction Meta:</h3>
+                        <p>Error: {data.meta.err}</p>
+                        <p>Fee: {data.meta.fee?.__bigint?.toString()}</p>
+                        <p>Log Messages: {data.meta.logMessages}</p>
+                        <p>Post Balances: {data.meta.postBalances?.__bigint?.toString()}</p>
+                        <p>Pre Balances : {data.meta.preBalances?.__bigint?.toString()}</p>
```
If any of these often come up undefined, like `meta.err`, we can just omit them from the rendering for now.

buffalojoec commented on src/fetch.ts:
```diff
@@ -20,3 +20,27 @@ export async function gql(...params: Parameters<RpcGraphQL['query']>) {
     });
 }
+
+export async function _fetch(url:string, body: any){
```
Just reiterating my request from earlier to roll a shared `post` function.

https://github.com/buffalojoec/graphql-wallet/pull/7#discussion_r1598862549

Hrushi20 committed:
lint, simple error handling on server, shared post function

buffalojoec commented on src/fetch.ts:
```diff
@@ -2,17 +2,28 @@ import type { RpcGraphQL } from '@solana/rpc-graphql';
+    return await postRequest('/api/graphql', { source, variableValues });
+}
+
+export async function _fetch(url: string, body: object) {
+    return await postRequest(url, body);
+}
```
There's no reason for this function anymore, since the app can now just call `postRequest` directly.

buffalojoec commented on src/fetch.ts:
```diff
@@ -2,17 +2,28 @@ import type { RpcGraphQL } from '@solana/rpc-graphql';
+    return await postRequest(url, body);
+}
+
+async function postRequest(url: string, body: object) {
```
```suggestion
async function postRequest(endpoint: string, body: object) {
```

buffalojoec commented on src/fetch.ts:
```diff
@@ -2,17 +2,28 @@ import type { RpcGraphQL } from '@solana/rpc-graphql';
         body: JSON.stringify({
-            source,
-            variableValues,
+            ...body,
```
I think you can just do `JSON.stringify(body)` here, can't you?

Hrushi20 replied:
By doing that we'll get the following payload as body:

```json
"body": {
  "body": {"source":"abc", "variableValues": "xyz"}
}
```

buffalojoec replied:
```ts
const body = { key: 'value' };
const stringified = JSON.stringify(body);
console.log(stringified);
```

```
"{"key":"value"}"
```

https://www.typescriptlang.org/play/?#code/MYewdgzgLgBARiAJgTxgXhgbxgawKbIBcMA5AG4CGANgK54kwC+A3AFCiSzQBOAlmAHNeAM155E6GACkAygHkAcgDoe-IcOQAKBCgCUbDhBBU8SqiAGbVgkWMT6gA

Hrushi20 replied:
Oh my bad. I didn't remove the `{}` in JSON.stringify. Hence I got that error. Thanks.

Hrushi20 replied:
Fixed it.

buffalojoec commented on src/fetch.ts:
```diff
@@ -2,17 +2,28 @@ import type { RpcGraphQL } from '@solana/rpc-graphql';
         method: 'POST',
     });
     const text = await res.text();
+    if (text === 'SERVER ERROR') {
+        console.error('Server Error. Check Server logs');
+        return;
+    }
```
You can just check for the `200` success code here.

buffalojoec commented on server.ts:
```diff
@@ -33,6 +34,20 @@ app.post('/api/graphql', async (req, res) => {
+
+        let results = await rpc.getSignaturesForAddress(address).send();
+        const signatures = results.map(result => result.signature );
+        console.log(signatures)
```
Let's not log the signatures on the server.

buffalojoec reviewed (commented):
Getting close! Thanks for the continued effort. Just a few last things to clean up.

Hrushi20 committed:
remove _fetch func

Hrushi20 committed:
Fix postRequest json stringify

buffalojoec reviewed (approved):
![ship it](https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExemZpanI4eGdoeHFhdTNmOTA2YW96Y3RoZ2VwcmxhdWxtM3M0d3Z0biZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/Hw8vYF4DNRCKY/giphy.gif)
